scene/3d/cpu_particles_3d.cpp

```diff
@@ -663,10 +663,19 @@ void CPUParticles3D::_particles_process(double p_delta) {
 
 	double system_phase = time / lifetime;
 
+	// Pre-resize arrays to improve performance.
+	Array cpu_particles;
+	cpu_particles.resize(pcount);
+
+	// Size is not known in advance, so we can't resize it beforehand.
+	Array cpu_particles_expired;
+
 	for (int i = 0; i < pcount; i++) {
 		Particle &p = parray[i];
+		Ref<CPUParticle3D> cpu_particle = memnew(CPUParticle3D);
 
```

Comment:
If this is using dynamic allocation for each particle within a particle system per frame, then it is a _really really_ bad idea. Particles should usually be super lightweight data orientated. If they absolutely have to be objects they should be pooled. But really they should just be data imo.

Reply:
There's already a Particle struct, but I can't emit it to scripting using a signal.

Reply:
> Can the signal not come from the particle system, rather than the particle?

This is already what happens, but the issue is passing the data to the signal in a way that is compatible with the scripting API. GDScript does not have a `struct` type, so type-safe data must be passed via custom objects.

We could emit an array of dictionaries in the signal, but it's not strongly typed. This means there's no autocompletion and errors will only occur at run-time in scripts (instead of compile-time).

Reply:
Making the particle system super inefficient to work around bugs / deficiencies in the scripting seems a case of the tail wagging the dog. Imo it is important the particle system works efficiently, then problems in gdscript can be addressed in further PRs.

Reply:
I've pushed a second commit that switches to a `Dictionary[]` return type.